(#2067) add support for `managed-dependencies`

cprice404 opened the pull request:
FOR REVIEW: this PR is just up for early review.  There is still some cleanup to do, and additional tests to add.  Also, I'm planning on tackling the "parent project" work somewhat separately, but it might make sense to hold off on this one until that part is at least somewhat proven out.  I'd appreciate any feedback on this PR in its current state, though.

---------------------------------------------------------------------------------------

This commit provides initial support for `managed-dependencies`,
where dependency version numbers may be specified in a separate
section called `managed-dependencies`, and those version numbers
will be used for any deps in the main `dependencies` section
that do not explicitly specify a version number.

This is a precursor to being able to specify a "parent" project
that could be used to consolidate version numbers of common
dependencies across a large number of libraries.

cprice404 commented on leiningen-core/src/leiningen/core/classpath.clj:
```diff
@@ -486,6 +493,19 @@
             (extract-native-dep! native-path file native-prefix))))))
 
+(defn resolve-managed-dependencies
+  "TODO"
```
Need to fix this

cprice404 commented on leiningen-core/src/leiningen/core/classpath.clj:
```diff
@@ -486,6 +493,19 @@
+      (extract-native-dependencies project jars dependencies-tree))
+    jars))
+
 (defn resolve-dependencies
```
I left this in because I didn't want to break API, but I am wondering if it should be marked as deprecated or something since it should no longer be used internally?

hyPiRion replied:
Yeah, I think we can add in `^:deprecated` to this one.

I looked over at [crossclj](https://crossclj.info/fun/leiningen.core.classpath/resolve-dependencies.html) for uses, and it seems to be used outside of lein in a couple of plugins. So I think that, to avoid breakage with pluginsm, it would be beneficial to replace `nil` with `:managed-dependencies` here if `(= dependencies-key :dependencies)`.

cprice404 commented on leiningen-core/src/leiningen/core/classpath.clj:
```diff
@@ -494,23 +514,25 @@
+(defn managed-dependency-hierarchy
+  "TODO"
+  [dependencies-key managed-dependencies-key project & options]
+  ;; TODO: explain private call
```
There is a new fn in `pomegranate` `0.3.1` that I didn't realize that I would need to call in this PR, so it's marked as private in `pomegranate`.  I can open a PR against `pomegranate` to try to get it switch to public, but I'm worried that it might take a long time to get another `pomegranate` release with that change.

hyPiRion replied:
It's fine to use a private one as we have full control of which pomegranate version we use. It would be beneficial to get it changed though.

cprice404 commented on test_projects/managed-deps/project.clj:
```diff
@@ -0,0 +1,12 @@
+
+  :dependencies [[org.clojure/clojure]
+                 [rome/rome nil]
+                 [ring]])
```
I'd like to add several more things to this file, such as deps that have exclusions, classifiers, etc.

cprice404 commented on test/leiningen/test/deps.clj:
```diff
@@ -128,3 +127,11 @@
+      (delete-file-recursively (m2-dir n v) :silently))
+    (deps managed-deps-project)
+    (doseq [[n v] managed-deps]
+      (is (.exists (m2-dir n v)) (str n " was not downloaded.")))))
```
Would like to expand this test a bit further, along with the project file that it uses.

cprice404 commented:
@hyPiRion if you get a chance at some point, I'd be interested to hear your thoughts on this, and input on next steps, etc.

hyPiRion commented:
@cprice404, thanks! And sure thing, I'll get a look whenever I have time.

cprice404 commented:
Travis failures all seem to be issues with downloading the clojure 1.7 jar from "foo.com".  This wasn't happening to me locally, guessing some of the tests are stubbed out to try not to go out to the internet and download artifacts from maven central, and that for these tests, the fact that I changed the clojure version from 1.3 to 1.7 means that the CI system doesn't already have the jars in place?

Will try clearing my local m2 cache and seeing if I can repro locally.

cprice404 commented:
Was able to repro travis failure locally, squashed up a commit that should fix.

cprice404 commented:
@hyPiRion ping just to make sure this didn't slip through the cracks.  If you just haven't had time yet, no worries, I totally understand!  The life of an OSS project maintainer :)

hyPiRion commented:
@cprice404 No worries, this is still on my list. I've been surprisingly busy the last two-three weeks, so unfortunately I've not been able to look at pull requests.

That seems to be over though, and I did some small patches last evening. Hopefully I'll be able to go through these during this week =)

cprice404 commented:
@hyPiRion excellent, thank you.

cprice404 commented:
@hyPiRion any chance to take a peek at this yet?

hyPiRion commented:
Alright, got around to look at this one. The initial implementation here looks good, I think.

I the only thing that should be fixed is to avoid breakage of existing plugins that use `resolve-dependencies`, as they probably want to resolve dependencies. That would break if managed-dependencies-key is set to nil rather than `:managed-dependencies`.

cprice404 commented:
@hyPiRion thanks very much!  I will look into the change you suggested.  Am also still soliciting another review from a co-worker, and then will start looking into the work to add support for the parent project file.

cprice404 commented:
OK I pushed up some commits addressing the issue pointed out in review, cleaning up some TODOs, and adding better test coverage.  I think this PR is reasonably merge-able at this point, but it's not very useful w/o the additional work to add support for parent projects.  I'm planning to start working on that soon, and probably just add it as additional commits on this PR.

hyPiRion commented:
The implementation looks and seems good, but I haven't gone through the test cases. When I've looked over them I'll merge this one in. =)

senior commented:
@cprice404 I use `sample.project.clj` pretty heavily when making a `project.clj` change. Any reason not to update that so users know it's there and how to use it?

cprice404 commented:
@senior no, that is a good suggestion.  I'll add something there and also take a poke around and see if there are any other docs that I need to mention this in.

cprice404 commented:
@senior I added a commit with an example and some docs.

cprice404 commented:
Also @senior did you look over my test coverage?  Anything you can think of that I'm not covering sufficiently?

senior commented on doc/MANAGED_DEPS.md:
```diff
@@ -0,0 +1,104 @@
+The `:managed-dependencies` section of your `project.clj` file is just like the
+regular `:dependencies` section, with two exceptions:
+
+1. It does not actually introduce any dependencies to your project.  It only says,
```
I think of it as, when a version is not specified, use this one. Reading this makes me think that it's more of an override.

senior commented on doc/MANAGED_DEPS.md:
````diff
@@ -0,0 +1,104 @@
+will allow you to define a single "parent" project that is inherited by multiple
+"child" projects; e.g.:
+
+```
````
Above you're denoting the block as clojure, here it's not

senior commented on test/leiningen/test/deps.clj:
```diff
@@ -128,3 +129,62 @@
+                         merged-deps)]
+    ;; make sure the sample data has some unmanaged deps, some unused managed deps,
+    ;; and some classified deps, for completeness
+    (is (not (empty? versioned-unmanaged-deps)))
```
seq is preferred over (not (empty?...))

cprice404 replied:
`seq?` doesn't have the same semantics.  The `remove` above may return `'()`, which evaluates to `true` with `seq?`, which makes the assertion invalid.

cprice404 replied:
oh, `seq`, not `seq?`.  That seems significantly less readable to me but I'll change it.

senior commented:
I think there might be some changes needed to the [pom generation code](https://github.com/technomancy/leiningen/blob/master/src/leiningen/pom.clj#L333). I'm not sure what the right behavior is. Probably at a minimum merging the managed dependencies in with the regular ones before generating the file

senior commented on test_projects/managed-deps/project.clj:
```diff
@@ -0,0 +1,18 @@
+                 [rome ~(str "0." "9")]
+                 [ring/ring "1.0.0"]
+                 [ring/ring-codec "1.0.1"]
+                 [commons-math/commons-math "1.2" :classifier "sources"]
```
In the docs you talk about managed deps being particularly useful for transitive dependencies. Is that tested here?

cprice404 commented:
@senior I added a commit that addresses all of your feedback *except* for the pom generation thing.  Still working on that, but if you wanted to confirm that this commit addresses the rest of it that'd be swell.

senior commented:
The changes look good but the managed-deps tests are failing

cprice404 commented:
derp.  squashed a fix that will hopefully get travis passing.

cprice404 commented:
@senior OK I pushed up another commit that adds `dependencyManagement` section to the pom.  Good catch on that.  I think I've addressed all of your feedback.

senior commented:
:+1:

cprice404 commented:
@hyPiRion : @senior helped me review test cases and found some things that I've since fixed up.  I'll sit on this for a few days in case you have time / interest to take another look at it, but I'll probably merge it soon-ish since it seemed like you were mostly +1 on it.

And, TBH, if you did end up with any bandwidth for PR review, I'd love to see if you have any thoughts on #2164 since that one hasn't had as many eyes on it.  Thanks!

cprice404 commented:
OK, went ahead and merge this since I think it's been reviewed pretty thoroughly.  @hyPiRion I'll keep an eye out for related bugs and try to make sure I address anything if it comes up.  I'm going to solicit some review from my co-workers on #2164, and once that is merged we will wait for the next lein release so that we can start getting some mileage on the `lein-parent` plugin!

cprice404 commented:
looks like this broke the build.  I'll start working on a fix right away.

hyPiRion commented:
For reference: #2170 is the actual issue here – the build failure is completely unrelated to this PR.
